controllers/topology/internal/extensions/patches/api/interface.go

```diff
@@ -0,0 +1,149 @@
+type GenerateRequest struct {
+	// Variables is a name/value map containing variables.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Items contains the list of templates to generate patches for.
+	Items []*GenerateRequestTemplate
+}
+
+// GenerateRequestTemplate defines one of the ClusterClass templates to generate patches for.
+type GenerateRequestTemplate struct {
+	// TemplateID identifies a template to generate patches for;
+	// the same TemplateID must be used when specifying to which template a generated patch should be applied to.
+	TemplateID TemplateID
+
+	// Variables is a name/value map containing variables specifically for the current template.
+	// For example some builtin variables like MachineDeployment replicas and version are context-sensitive
+	// and thus are only added to templates for MachineDeployments and with values which correspond to the
+	// current MachineDeployment.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Template contains the template.
+	Template apiextensionsv1.JSON
+}
+
+// TemplateID identifies one of the ClusterClass templates to generate patches for;
+// the same TemplateID must be used when specifying where a generated patch should apply to.
+type TemplateID struct {
+	// APIVersion of the current template.
+	APIVersion string
+
+	// Kind of the current template.
+	Kind string
+
+	// TargetType defines where the template is used.
+	TargetType TargetType
+
+	// MachineDeployment specifies the MachineDeployment in which the template is used.
+	// This field is only set if the template is used in the context of a MachineDeployment.
+	MachineDeployment MachineDeploymentID
+}
+
+// MachineDeploymentID specifies the MachineDeployment in which the template is used.
+type MachineDeploymentID struct {
+	// TopologyName is the name of the MachineDeploymentTopology.
+	TopologyName string
+
+	// Class is the name of the MachineDeploymentClass.
+	Class string
+}
+
+// TargetType define the type for target types enum.
+type TargetType int32
+
+const (
+	// InfrastructureClusterTemplateTargetType identifies a template for the InfrastructureCluster object.
+	InfrastructureClusterTemplateTargetType TargetType = iota
+
+	// ControlPlaneTemplateTargetType identifies a template for the ControlPlane object.
+	ControlPlaneTemplateTargetType
+
+	// ControlPlaneInfrastructureMachineTemplateTargetType identifies a template for the InfrastructureMachines to be used for the ControlPlane object.
+	ControlPlaneInfrastructureMachineTemplateTargetType
+
+	// MachineDeploymentBootstrapConfigTemplateTargetType identifies a template for the BootstrapConfig to be used for a MachineDeployment object.
+	MachineDeploymentBootstrapConfigTemplateTargetType
+
+	// MachineDeploymentInfrastructureMachineTemplateTargetType identifies a template for the InfrastructureMachines to be used for a MachineDeployment object.
+	MachineDeploymentInfrastructureMachineTemplateTargetType
+)
```

Comment:
Why use iota instead of typed strings? The above isn't very portable to other programming languages

Reply:
Sounds reasonable

Reply:
FYI Using enum originate from the protobuf experiment, but no objection to change

Reply:
Okay let's change it for now. Potentially we change it back when we introduce external patches.